lib/core/waitForChainlockedHeightFactory.js

```diff
@@ -0,0 +1,40 @@
+const MissingChainLockError = require('./errors/MissingChainLockError');
+const wait = require('../util/wait');
+
+/**
+ *
+ * @param {LatestCoreChainLock} latestCoreChainLock
+
+ * @return {waitForChainlockedHeight}
+ */
+function waitForChainlockedHeightFactory(
+  latestCoreChainLock,
+) {
+  /**
+   * @typedef waitForChainlockedHeight
+   * @param {number} coreHeight
+   *
+   * @return {Promise<void>}
+   */
+  async function waitForChainlockedHeight(coreHeight) {
+    // ChainLock is required to get finalized SML that won't be reorged
+    let chainLock = latestCoreChainLock.getChainLock();
+
+    if (!chainLock) {
+      throw new MissingChainLockError();
+    }
+
+    // Wait for core to be synced up to coreHeight
+    if (coreHeight > chainLock.height) {
```

Comment:
You can just listen for update event and resolve promise when you get the required height. It will be much better than loop.